pkg/trace/stats/client_stats_aggregator.go

```diff
@@ -0,0 +1,295 @@
+// If a single payload is received on a bucket, this Aggregator is a passthrough.
+// If two or more payloads collide, their counts will be aggregated into one bucket.
+// Multiple payloads will be sent:
+// - Original payloads with their distributions will be sent with counts zeroed.
+// - A single payload with the bucket aggregated counts will be sent.
+// This ensures that all counts will have at most one point per second per agent for a specific granularity.
+// While distributions are not tied to the agent.
+type ClientStatsAggregator struct {
+	In      chan pb.ClientStatsPayload
+	out     chan pb.StatsPayload
+	buckets map[uint64]*bucket // buckets used to aggregate client stats
+
+	flushTicker   *time.Ticker
+	oldestTs      uint64
+	agentEnv      string
+	agentHostname string
+
+	exit chan struct{}
+	done chan struct{}
+}
+
+// NewClientStatsAggregator initializes a new aggregator ready to be started
+func NewClientStatsAggregator(conf *config.AgentConfig, out chan pb.StatsPayload) *ClientStatsAggregator {
+	return &ClientStatsAggregator{
+		flushTicker:   time.NewTicker(time.Second),
+		In:            make(chan pb.ClientStatsPayload, 10),
+		buckets:       make(map[uint64]*bucket, 20),
+		out:           out,
+		agentEnv:      conf.DefaultEnv,
+		agentHostname: conf.Hostname,
+		oldestTs:      uint64(time.Now().Add(-oldestBucketStart).Unix()),
+		exit:          make(chan struct{}),
+		done:          make(chan struct{}),
+	}
+}
+
+// Start starts the aggregator.
+func (a *ClientStatsAggregator) Start() {
+	go func() {
+		defer watchdog.LogOnPanic()
+		for {
+			select {
+			case t := <-a.flushTicker.C:
+				a.flushOnTime(t)
+			case input := <-a.In:
+				a.add(time.Now(), input)
+			case <-a.exit:
+				a.flushAll()
+				close(a.done)
+				return
+			}
+		}
+	}()
+}
+
+// Stop stops the aggregator. Calling Stop twice will panic.
+func (a *ClientStatsAggregator) Stop() {
+	close(a.exit)
+	<-a.done
+}
+
+// flushOnTime flushes all buckets up to t, except the last one.
+func (a *ClientStatsAggregator) flushOnTime(t time.Time) {
+	flushTs := uint64(t.Unix()) - uint64(oldestBucketStart.Seconds())
+	for ts := a.oldestTs; ts < flushTs; ts++ {
+		if b, ok := a.buckets[ts]; ok {
+			for _, p := range b.flush() {
+				a.flush(p)
+			}
+			delete(a.buckets, ts)
+		}
+	}
+	a.oldestTs = flushTs
+}
+
+func (a *ClientStatsAggregator) flushAll() {
+	for _, b := range a.buckets {
+		for _, p := range b.flush() {
+			a.flush(p)
+		}
+	}
+}
+
+// getAggregationBucketTime returns unix time at which we aggregate the bucket.
+// We timeshift payloads older than a.oldestTs to a.oldestTs.
+// Payloads in the future are timeshifted to time.Now().Unix()
+func (a *ClientStatsAggregator) getAggregationBucketTime(t time.Time, bucketStart uint64) (uint64, bool) {
+	ts := bucketStart / 1e9 // conversion from nanosecond to second
+	if ts < a.oldestTs {
+		return a.oldestTs, true
+	}
+	now := uint64(t.Unix())
+	if ts > now {
+		return now, true
+	}
```

Comment:
In general, I think this struct is doing too much translation back and forth between `time.Time` and int types. This opens the door to potential bugs. Can we use `time.Time` everywhere except for the map key? If we need to truncate to a second, let's do so explicitly with time's `Truncate` function.